AK/URL.h

```diff
@@ -84,6 +84,9 @@ class URL {
     DeprecatedString path_segment_at_index(size_t index) const;
     size_t path_segment_count() const { return m_paths.size(); }
 
+    // A URL has an opaque path if its path is a URL path segment.
+    bool has_opaque_path() const { return path_segment_count() == 1; }
```

Comment:
I think it is quite confusing, but I believe `cannot_be_a_base_url` may just be equivalent to "has an opaque path" or at least, that's how we seem to treat it at the moment?). Also, is it correct to say that it has an opaque path if the path segment count is _exactly_ one?

Ref: https://github.com/SerenityOS/serenity/blob/851ac682efb893f206e0dab793d5480ee5473e58/AK/URLParser.cpp#L920

Reply:
Thanks for looking into this!

>  think it is quite confusing, but I believe cannot_be_a_base_url may just be equivalent to "has an opaque path" or at least, that's how we seem to treat it at the moment?).

Yeah I found calls to this method while reading the `LibWeb::URL` code and wondered if I could use that instead, but was deterred when I saw that anyone could set the value since it also had a setter. I added that method to `AK::URL` since it seemed to align with how the [spec ](https://url.spec.whatwg.org/#url-opaque-path) defined it.

I also feel that based off the name of `cannot_be_a_base_url `, one could see both `https://example.com/a` (which has an opaque path) and `https://example.com/a/b/c` (not an opaque path) as not being base URLs. Let me know what you think!

> Also, is it correct to say that it has an opaque path if the path segment count is exactly one?

The spec phrased it in a way that would mean their could be more than one path fragment (even though that's not the case since that goes against its definition). I was confused by this too, but thought it was safer to be consistent with the spec. I'll make a change for that.

Reply:
I found one relevant change to the spec, the old phrasing "cannot be a base URL" in certain places seem to have been renamed: https://github.com/whatwg/url/commit/fbaa03cb19ee5718953f5f6d179e0339e31f1ede

We have artifacts of that change in other locations in our code, for example: https://github.com/SerenityOS/serenity/blob/ec416563a896f188072bac6e86a3234888a8f231/Userland/Libraries/LibWeb/URL/URL.cpp#L282

But if you view: https://url.spec.whatwg.org/#dom-url-hostref-for-dom-url-host%E2%91%A0 you'll notice that the phrasing has been updated per that above commit.

When I get the time a little later, I'll try dig some more to understand. Seems we've become out of date with the spec at some stage, and we need to go through and fix up the spec comments / terminology at some stage to make this more clear.

Reply:
Do you think I should still include `is_opaque_path()` in `AK::URL` or maybe just have `LibWeb::URL::is_opaque_path()` call on `AK::URL::cannot_be_a_base_url()` for the time being?

Reply:
My guess is that we may need a new function for this, but would I need to spend some time reading the spec to properly understand what we should be doing here.

Reply:
Okay, I've found some time to look at this properly. I don't think this implementation of `has_opaque_path` is correct. It looks like this was originally changed in: https://github.com/whatwg/url/pull/655

Later on, this was refined in: https://github.com/whatwg/url/pull/736

For now, I think it's okay to use `cannot_be_a_base_url`. This explains some of the issues you were having with: https://github.com/SerenityOS/serenity/pull/21021#discussion_r1321260808

Probably we need to catch up with the spec a little bit for this all to make a little more sense. Trying to figure out at the moment what changes we need to be making.